File: backend/src/services/restaurant.service.ts
import {
  CreateRestaurantDto,
  DeleteRestaurantDto,
  FilterOperator,
  FilterOptions,
  FilterRelation,
  FilterRelations,
  GetRestaurantDto,
  GetRestaurantsDto,
  UpdateRestaurantDto,
} from '@dtos/restaurant.dto';
import { HttpException } from '@exceptions/http.exception';
import { Restaurant, Restaurants } from '@interfaces/restaurant.interface';
import { Status } from '@interfaces/status.interface';
import { User } from '@interfaces/user.interface';
import restaurantModel from '@models/restaurant.model';
import { filterOptionsToMongoQuery } from '@utils/utils';

export default class RestaurantService {
  private restaurants = restaurantModel;

  public async create(restaurantInfo: CreateRestaurantDto, user: User): Promise<Restaurant> {
    const restaurant: Restaurant = await this.restaurants.create({ ...restaurantInfo, ownerId: user._id });
    return restaurant;
  }

  public async getAll(dto: GetRestaurantsDto): Promise<Restaurants> {
    const restaurants: Restaurant[] = await this.restaurants
      .where({ status: Status.Active, ...filterOptionsToMongoQuery(dto.filterOptions) })
      .sort({ ...dto.sortOptions })
      .skip(dto.paginationOptions.pageNumber * dto.paginationOptions.itemsPerPage)
      .limit(dto.paginationOptions.itemsPerPage);
    return { restaurants: restaurants };
  }

  public async get(dto: GetRestaurantDto): Promise<Restaurant> {
    const restaurant: Restaurant = await this.restaurants.findOne({ _id: dto.id, status: Status.Active });
    if (!restaurant) {
      throw new HttpException(404, 'No restaurant with this id was found.');
    }
    return restaurant;
  }

  public async delete(dto: DeleteRestaurantDto): Promise<Restaurant> {
    const updatedRestaurant: Restaurant = await this.restaurants.findOneAndUpdate(
      { _id: dto.id, status: Status.Active },
      { status: Status.Disabled },
      { new: true },
    );
    if (!updatedRestaurant) {
      throw new HttpException(404, 'No restaurant with this id was found.');
    }

    return updatedRestaurant;
  }

  public async update(dto: UpdateRestaurantDto): Promise<Restaurant> {
    const updatedRestaurant: Restaurant = await this.restaurants.findOneAndUpdate(
      { _id: dto.id, status: Status.Active },
      { name: dto.name, address: dto.address },
      { new: true },
    );
    if (!updatedRestaurant) {
      throw new HttpException(404, 'No restaurant with this id was found.');
    }

    return updatedRestaurant;
  }
}

File: frontend/lib/services/restaurant.service.dart
import 'package:flutter/foundation.dart';
import 'package:foodz/api/api.client.dart';
import 'package:foodz/endpoints/restaurant.endpoint.dart';
import 'package:foodz/models/dtos/pagination.options.dto.dart';
import 'package:foodz/models/dtos/restaurant.dto.dart';
import 'package:foodz/models/restaurant.model.dart';
import 'package:foodz/services/exceptions/restaurant.service.exception.dart';
import 'package:foodz/util/value.range.dart';

class RestaurantService {
  Future<Restaurant> create(
      {@required String name, @required String address}) async {
    final createRestaurantDto =
        CreateRestaurantDto(name: name, address: address);

    final requestResult = await ApiClient().request<Restaurant>(
        RestaurantEndpoint.create(createRestaurantDto: createRestaurantDto)
            .config,
        Restaurant.fromJson);

    if (requestResult.isError) {
      throw RestaurantServiceException(requestResult);
    }

    return requestResult.data;
  }

  Future<List<Restaurant>> getAll(
      {String ownerId,
      ValueRange<double> ratingRange,
      int pageNumber,
      int itemsPerPage}) async {
    final getRestaurantsDto = GetRestaurantsDto(
        sortOptions: SortOptions(averageRating: SortCriteria.Descending),
        filterOptions: FilterOptions(
            ownerId: ownerId == null
                ? null
                : FilterRelations.single(
                    relation: FilterRelation(
                        filterOperator: FilterOperator.Equal, value: ownerId)),
            averageRating: ratingRange == null
                ? null
                : FilterRelations(relations: [
                    if (ratingRange != null && ratingRange.min != null)
                      FilterRelation(
                          filterOperator: FilterOperator.GeaterThanOrEqual,
                          value: ratingRange.min),
                    if (ratingRange != null && ratingRange.max != null)
                      FilterRelation(
                          filterOperator: FilterOperator.LessThanOrEqual,
                          value: ratingRange.max)
                  ])),
        paginationOptions: PaginationOptions(
            pageNumber: pageNumber, itemsPerPage: itemsPerPage));
    final requestResult = await ApiClient().request<Restaurants>(
        RestaurantEndpoint.getAll(getRestaurantsDto: getRestaurantsDto).config,
        Restaurants.fromJson);

    if (requestResult.isError) {
      throw RestaurantServiceException(requestResult);
    }

    return requestResult.data.restaurants;
  }

  Future<Restaurant> getInfo({@required String id}) async {
    final getRestaurantDto = GetRestaurantDto(id: id);
    final requestResult = await ApiClient().request<Restaurant>(
        RestaurantEndpoint.getInfo(getRestaurantDto: getRestaurantDto).config,
        Restaurant.fromJson);

    if (requestResult.isError) {
      throw RestaurantServiceException(requestResult);
    }

    return requestResult.data;
  }

  Future<Restaurant> update(
      {String restaurantId, String name, String address}) async {
    final updateRestaurantDto =
        UpdateRestaurantDto(id: restaurantId, name: name, address: address);
    final requestResult = await ApiClient().request<Restaurant>(
        RestaurantEndpoint.update(updateRestaurantDto: updateRestaurantDto)
            .config,
        Restaurant.fromJson);

    if (requestResult.isError) {
      throw RestaurantServiceException(requestResult);
    }

    return requestResult.data;
  }

  Future<Restaurant> delete({@required String restaurantId}) async {
    final deleteRestaurantDto = DeleteRestaurantDto(id: restaurantId);
    final requestResult = await ApiClient().request<Restaurant>(
        RestaurantEndpoint.delete(deleteRestaurantDto: deleteRestaurantDto)
            .config,
        Restaurant.fromJson);

    if (requestResult.isError) {
      throw RestaurantServiceException(requestResult);
    }

    return requestResult.data;
  }
}
